In [8]:
from nltk.corpus import wordnet

## 1.
 Найти все значения (синсеты) для лексемы *mole* 

In [9]:
mole = wordnet.synsets('mole')
for ss in mole:
    print(ss, ss.definition())

Synset('gram_molecule.n.01') the molecular weight of a substance expressed in grams; the basic unit of amount of substance adopted under the Systeme International d'Unites
Synset('counterspy.n.01') a spy who works against enemy espionage
Synset('mole.n.03') spicy sauce often containing chocolate
Synset('mole.n.04') a small congenital pigmented spot on the skin
Synset('breakwater.n.01') a protective structure of stone or concrete; extends from shore into the water to prevent a beach from washing away
Synset('mole.n.06') small velvety-furred burrowing mammal having small eyes and fossorial forefeet


## 2.
Найти определение для лексемы mole в значении (а) "родинка" и в значении (b) "крот"

In [10]:
# родинка
mole[3], mole[3].definition()

(Synset('mole.n.04'), 'a small congenital pigmented spot on the skin')

In [12]:
# крот
mole[5], mole[5].definition()

(Synset('mole.n.06'),
 'small velvety-furred burrowing mammal having small eyes and fossorial forefeet')

## 3.
Найдите два произвольных контекста для слова plant в значениях (a) "родинка" и (b) "крот"; продемонстрируйте на них действие алгоритма Леска для разрешения неоднозначности

In [13]:
spot = 'I just came from the doctor and the mole is absolutely fine.'
animal = 'When a mole digs a tunnel where does the dirt go?'

In [14]:
spot_tokens = [i.strip('.?') for i in spot.split()]
animal_tokens = [i.strip('.?') for i in animal.split()]

In [15]:
spot_tokens, animal_tokens

(['I',
  'just',
  'came',
  'from',
  'the',
  'doctor',
  'and',
  'the',
  'mole',
  'is',
  'absolutely',
  'fine'],
 ['When',
  'a',
  'mole',
  'digs',
  'a',
  'tunnel',
  'where',
  'does',
  'the',
  'dirt',
  'go'])

In [17]:
from nltk.wsd import lesk
print(lesk(spot_tokens, 'mole').definition())

a protective structure of stone or concrete; extends from shore into the water to prevent a beach from washing away


In [18]:
print(lesk(animal_tokens, 'mole').definition())

a small congenital pigmented spot on the skin


Алгоритм Леска ошибся в обоих случаях. 

## 4.
Найдите гиперонимы для значения (a) и гиперонимы для значения (b)

In [20]:
for ss in mole[3].hypernyms():
    print(ss, ss.definition())

Synset('blemish.n.01') a mark or flaw that spoils the appearance of something (especially on a person's body)


In [21]:
for ss in mole[5].hypernyms():
    print(ss, ss.definition())

Synset('insectivore.n.01') small insect-eating mainly nocturnal terrestrial or fossorial mammals


## 5.
Вычислите наименьшее расстояние между значением mole "родинка" и значениями лексемы face, а также mole "крот" и значениями лексемы digger. 

In [22]:
def get_dist_sim(ss1, lexeme):
    distances = []
    similarities = []
    for ss2 in lexeme:
        dist = ss1.shortest_path_distance(ss2)
        if dist is not None:
            distances.append(dist)
            sim = ss1.path_similarity(ss2)
            similarities.append(sim)
    return distances, similarities

In [41]:
face = wordnet.synsets('face')
digger = wordnet.synsets('digger')

In [42]:
# min d(mole "родинка", face)
dist1 = get_dist_sim(mole[3], face)[0]
print('min d(mole: "родинка", face): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(face[dist1.index(min(dist1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# min d(mole "родинка", digger)
dist2 = get_dist_sim(mole[3], digger)[0]
print('min d(mole: "родинка", digger): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(digger[dist2.index(min(dist2))].definition()))

# min d(mole: "крот", face)
dist3 = get_dist_sim(mole[5], face)[0]
print('min d(mole: "крот", face): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(face[dist3.index(min(dist3))].definition()))

# min d(mole: "крот", digger)
dist4 = get_dist_sim(mole[5], digger)[0]
print('min d(mole: "крот", digger): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(digger[dist4.index(min(dist4))].definition()))

print('min (d(mole: "родинка", face), d(mole: "родинка", digger)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(mole: "крот", face), d(mole: "крот", digger)): {}'.format(min(min(dist3), min(dist4))))

min d(mole: "родинка", face): 3
closest lemma definition: the general outward appearance of something

min d(mole: "родинка", digger): 14
closest lemma definition: a laborer who digs

min d(mole: "крот", face): 9
closest lemma definition: a part of a person that is used to refer to a person

min d(mole: "крот", digger): 13
closest lemma definition: a laborer who digs

min (d(mole: "родинка", face), d(mole: "родинка", digger)): 3
min (d(mole: "крот", face), d(mole: "крот", digger)): 9


In [43]:
# min d(mole "родинка", face)
dist1 = get_dist_sim(mole[3], face)[1]
print('min d(mole: "родинка", face): {}'.format(min(dist1)))
print('closest lemma definition: {}\n'.format(face[dist1.index(min(dist1))].definition()))
# Правда, здесь выводится только одно ближайшее значение. Их может быть несколько.

# min d(mole "родинка", digger)
dist2 = get_dist_sim(mole[3], digger)[1]
print('min d(mole: "родинка", digger): {}'.format(min(dist2)))
print('closest lemma definition: {}\n'.format(digger[dist2.index(min(dist2))].definition()))

# min d(mole: "крот", face)
dist3 = get_dist_sim(mole[5], face)[1]
print('min d(mole: "крот", face): {}'.format(min(dist3)))
print('closest lemma definition: {}\n'.format(face[dist3.index(min(dist3))].definition()))

# min d(mole: "крот", digger)
dist4 = get_dist_sim(mole[5], digger)[1]
print('min d(mole: "крот", digger): {}'.format(min(dist4)))
print('closest lemma definition: {}\n'.format(digger[dist4.index(min(dist4))].definition()))

print('min (d(mole: "родинка", face), d(mole: "родинка", digger)): {}'.format(min(min(dist1), min(dist2))))
print('min (d(mole: "крот", face), d(mole: "крот", digger)): {}'.format(min(min(dist3), min(dist4))))

min d(mole: "родинка", face): 0.06666666666666667
closest lemma definition: a surface forming part of the outside of an object

min d(mole: "родинка", digger): 0.06666666666666667
closest lemma definition: a laborer who digs

min d(mole: "крот", face): 0.047619047619047616
closest lemma definition: a specific size and style of type within a type family

min d(mole: "крот", digger): 0.06666666666666667
closest lemma definition: a machine for excavating

min (d(mole: "родинка", face), d(mole: "родинка", digger)): 0.06666666666666667
min (d(mole: "крот", face), d(mole: "крот", digger)): 0.047619047619047616


По расстоянию "родинка" и "крот" ближе к "face", чем к "digger". В случае родинки - хорошо, в случае крота - скорее странно. По сходству путей на основе кратчайшего расстояния все примерно так же, однако расстояние от "родинки" до face и до digger одинаковое,что тоже странно. 

## 6.
Вычислить двумя разными способами расстояние:
d(mole: "крот", sow) и d(organism, whole)  
sow - самка крота  (согласно википедии)   
Я буду использовать Leacock-Chodorow Similarity и Wu-Palmer Similarity,

In [47]:
sow = wordnet.synsets("sow")
for ss in sow:
    print(ss, ss.definition())

Synset('sow.n.01') an adult female hog
Synset('sow.v.01') place (seeds) in or on the ground for future growth
Synset('sow.v.02') introduce into an environment
Synset('inseminate.v.01') place seeds in or on (the ground)


А здесь это самка свиньи. Ну почти.

In [50]:
#Leacock-Chodorow Similarity
print(mole[5].lch_similarity(sow[0]))

#Wu-Palmer Similarity
print(mole[5].wup_similarity(sow[0]))

1.6916760106710724
0.7857142857142857


In [48]:
organism = wordnet.synsets('organism')
for ss in organism:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('organism.n.01') a living thing that has (or can develop) the ability to act or function independently
Synset('organism.n.01') []
Synset('organism.n.02') a system considered analogous in structure or function to a living body
Synset('organism.n.02') ['the social organism']


In [49]:
whole = wordnet.synsets('whole', 'n')
for ss in whole:
    print(ss, ss.definition())
    print(ss, ss.examples())

Synset('whole.n.01') all of something including all its component elements or parts
Synset('whole.n.01') ['Europe considered as a whole', 'the whole of American literature']
Synset('whole.n.02') an assemblage of parts that is regarded as a single entity
Synset('whole.n.02') ['how big is that part compared to the whole?', 'the team is a unit']


In [52]:
for ss1 in organism:
    print(ss1.definition())
    for ss2 in whole:
        print(ss2.definition())

        #Leacock-Chodorow Similarity
        print(ss1.lch_similarity(ss2))

        #Wu-Palmer Similarity
        print(ss1.wup_similarity(ss2))


a living thing that has (or can develop) the ability to act or function independently
all of something including all its component elements or parts
1.072636802264849
0.14285714285714285
an assemblage of parts that is regarded as a single entity
2.538973871058276
0.8
a system considered analogous in structure or function to a living body
all of something including all its component elements or parts
1.3350010667323402
0.3076923076923077
an assemblage of parts that is regarded as a single entity
1.55814461804655
0.2222222222222222


|                                   | Leacock-Chodorow   | Wu-Palmer           |
|-----------------------------------|--------------------|---------------------|
| mole/sow                          | 1.6916760106710724 | 0.7857142857142857  |
| a living thing / all of something | 1.072636802264849  | 0.14285714285714285 |
| a living thing / an assemblage    | 2.538973871058276  | 0.8                 |
| a system / all of something       | 1.3350010667323402 | 0.3076923076923077  |
| a system / an assemblage          | 1.55814461804655   | 0.2222222222222222  |

+ И по метрике Leacock-Chodorow, и по метрике Wu-Palmer между mole и sow почти самое большое расстояние среди всех рассмотренных пар. (Дальше друг от друга только пара a living thing / an assemblage). Что неудивительно, учитывая, что в wordnet sow относится к свинье, а не к кроту. 
+ По моим интуитивным представлениям, если говорить про organism и whole, кажется, что наиболее близкими друг к другу должны быть слова в парах a system / all of something и a system / an assemblage. Но это не совсем так.
+ Между словами в паре a system / an assemblage действительно по обеим метрикам расстояние меньше, чем между словами в паре a living thing / an assemblage.
+ Однако пара a living thing / all of something по обем метрикам превосходит пару a system / all of something. 